### PR TITLE
Dockerize NUC and Laptop Setup

### PR DESCRIPTION
**Pull request by peterdavidfagan**
**Overview**
This pull request adds an scripts to automate the setup of the R2D2 application on both control and client devices. In particular it adds scripts under `r2d2/setup` for both the nuc and the laptop devices. These scripts leverage docker to containerize application code, the remainder of the setup scripts include host specific setup (e.g. rt patch of kernel, setting static ip addresses). 

**Contributors**
Thanks to @arehman1806 for assisting in the laptop setup and debuging the automated setup for the laptop. Thanks to @jrome5 for helping to debug apk app installation and issues relating to network latency from the RAD labs NUC device. Finally thank you to @AlexanderKhazatsky whose guides this work is based on, the goal of this pull request is to contribute an automated setup guide based on the guide defined by @AlexanderKhazatsky, hopefully these changes accomplish this.

### CODE REVIEW

File: scripts/setup/intro.txt [hunk 1]
```diff
@@ -0,0 +1,50 @@
+ .----------------.  .----------------.  .----------------.  .----------------.
```
Comment (blake-wulfe-tri) [inline scripts/setup/intro.txt:1]: nice!

**Comment by peterdavidfagan** [comment]
Hi @AlexanderKhazatsky and @blake-wulfe-tri, 

These changes are validated on our setup and should now be ready to 🚢.

Shipping Time

File: .docker/README.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# Overview
```
Comment (blake-wulfe-tri) [inline .docker/README.md:1]: helpful readme, thanks for adding!

File: .docker/README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+* [Dockerfile Reference](https://docs.docker.com/engine/reference/builder/)
+* [Docker Compose Overview](https://docs.docker.com/compose/)
+
+# NUC Setup
+In order to set up the control server on your NUC run `sudo ./nuc_setup.sh` from this [directory](https://github.com/AlexanderKhazatsky/R2D2/tree/main/scripts/setup). Running through all the steps in this script will install host system dependencies and ensure the control server runs automatically in a docker container each time your machine is booted.
+
+Further details on the steps in README.md in the `scripts/setup` directory.
+
+# Laptop Setup  
+In order to setup the nuc run `sudo ./laptop_setup.sh` from this [directory](https://github.com/AlexanderKhazatsky/R2D2/tree/main/scripts/setup). Running through all the steps in this script will install host system dependencies and ensure the user client can be run in a docker container.
```
Comment (blake-wulfe-tri) [inline .docker/README.md:25]: is this "setup the nuc" a typo?

**Review by blake-wulfe-tri — APPROVED** [review]
lgtm! I think automating / improving the automation of a lot of this will be a huge help to people, thanks for doing it! I will likely try this stuff out next week; we shouldn't block on that, though.

File: .docker/README.md (hunk 3 above)
Comment (peterdavidfagan) [reply .docker/README.md:25]: Thanks for catching this @blake-wulfe-tri, definitely a typo, I'll push a fix now.

File: .docker/README.md (hunk 2 above)
Comment (peterdavidfagan) [reply .docker/README.md:1]: Thanks for the suggestion to add this.

File: .docker/README.md (hunk 3 above)
Comment (peterdavidfagan) [reply .docker/README.md]: Closing as now resolved in [551359d](https://github.com/AlexanderKhazatsky/R2D2/pull/5/commits/551359d9cc97b7cef9e4396bc31d6fcc7b09bf3b).

**Comment by peterdavidfagan** [comment]
> lgtm! I think automating / improving the automation of a lot of this will be a huge help to people, thanks for doing it! I will likely try this stuff out next week; we shouldn't block on that, though.

Thanks for the review.

**Comment by blake-wulfe-tri** [comment]
I ran through part of the laptop setup. I'm trying to install it not on a laptop or nuc, but on a separate computer for the purposes of doing some data manipulation (so the laptop setup didn't work for me ultimately, but that's due to a lack of nuc, devices, etc for setup). It would be nice in a future PR if we could add a docker setup option for that (non-laptop, non-nuc, minimal requirements to build the docker image that installs r2d2 and deps).

**Comment by peterdavidfagan** [comment]
>  for the purposes of doing some data manipulation

Ah this is a good point. Our lab is only starting data collection so the lack of support for this is indicative of us only catching up on the project. 

>  It would be nice in a future PR if we could add a docker setup option for that

I believe it should be possible to spin a container with the R2D2 dependencies installed and run an interactive session with the current setup through running:

`docker compose -f docker-compose-laptop.yaml run --entrypoint=/bin/bash setup_laptop`

Do you think I should add another script that runs similar commands to spin the container and/or open an interactive session?

**Comment by AlexanderKhazatsky** [comment]
Hey Peter! Sorry have been tentative to pull this in since all labs are set up and it seemed you were still iterating rapidly. If you're very confident on the safety of this merge, ill merge it in! Thoughts? I'm pretty unfamiliar with docker so my ability to assess on my own is pretty limited.

**Comment by peterdavidfagan** [comment]
Hey Sasha, 

No worries at all and this is completely understandable. I think it makes sense to hold off given all labs are set up and this may cause some confusion. In the meantime, I'll make changes to improve the PR overtime, it is stable on our setup but there are likely some improvements that can be made (e.g. including docker compose files for data manipulation and upload). Maybe after some time when the initial data collection phase is complete and the project is published it may make sense to incorporate the docker setup, I agree that it may introduce some confusion to incorporate these changes now (for instance dependencies are incorporated as gitsubmodules which is different to the setup script).

Also good news from our lab this week, we got the Ulanzi camera mounts so we plan to resume data collection as we can now make alterations more easily (adjust desk height, angle) without having to recalibrate.

**Comment by AlexanderKhazatsky** [comment]
Great insight! Yes let's pause this for now and return to it  loser to publication, as this would be an absolutely awesome feature to release the dataset with.

Also, great to hear that your setup is complete!! Great work!!!

**Pull request merged**